### DISCUSSION TITLE
Health of the project

### DISCUSSION THREAD

**Discussion by TSnatch72** [florisboard/florisboard#2284] [General]
It's very evident that the project has been in a serious stallment in the last year. However, this is not only one of my favorite android projects ever but i also think is one of the most important if we ever want to have a fully de-googled yet fully functional, no compromises FOSS android ecosystem. 
So i want to ask: What's the project needing the most right now? What can we do to contribute? is the project lacking proper financing needed to keep it running? Is it purely lack of developers and commits? Does the maintainer not have the time and resources to properly manage it? is the project in need of volunteers helping to maintain it?

Please let's not let this awesome project fall into abandonment. 🙏🏽

**Comment by patrickgold** [comment]
Hey,

I'm Patrick the lead maintainer of this project and also the reason why this project has kinda stalled so much lately. Below I wanna elaborate why this has happened and also answer your questions.

> So i want to ask: What's the project needing the most right now?

Tbh at this stage and size of the project it has become more than evident that I am not able to manage it alone anymore, especially as I also need to manage personal responsibilities that are outside of this project's scope. I really love this project though and it would be really sad to see this project fall into abandonment just because I can't get my sh*t together, so I really need to evaluate to what extend this project needs volunteer developer resources to be able to keep up with the current amount of issues.

On the technical side the biggest hurdle is suggestions/NLP, while they kinda work on PC on mobile the performance is really bad and if I then try to optimize it I kinda make the suggestions garbage again, so in this regard I'm kinda completely stuck in a loop atm without a solution at the horizon. I lack expertise on this topic which makes matters worse and which kinda demotivates a bit.

On the development side my main problem is that I can't work on 5 things at once, so experimenting on NLP kinda prevents me from implementing/fixing other features, which again is evidence that developing this project alone is not doable anymore. There are some people who contributed awesome features but PR review takes time too which kinda just feeds into the time issue again.

> is the project lacking proper financing needed to keep it running?

It is not a financial problem at all, donations are a nice side product of this project, but my opinion on this is that throwing money on a problem doesn't make the problem go away and this project is not expensive to run at all from a financial standpoint.

**Reply by ray-holmium** [reply]
Patrick, thanks for sharing this update! I'm sure you've stated this elsewhere, but why did you settle on crafting the NLP yourself? Is there another library we could use, even as a stop-gap to take the pressure off you while you work on the ideal NLP?

**Reply by patrickgold** [reply]
As it must run fully offline no, i haven't really found any library for bridging NLP. One could say we could use the binary dictionaries of the original open-source Android LatinIME implementation but that code is very old, almost undocumented and thus not really understandable/usable at all.

It wouldn't nearly be as hard if I had more understanding of the statistical concepts behind NLP. There are a lot of resources online which kinda introduce you into the topic but no resource really explains it in a way in which all the small parts make sense together. Also sometimes it makes sense from a mathematical standpoint but then there's the issue of CPU constraints/memory constraints, so the combination is just really really hard to get somewhat right. And for a pleasant experience suggestions must be computed in max. 200ms else the lag is really noticeable.

What I also thought of using is offline machine learning models instead of statistical methods for NLP, however without utilizing existing Google services on the device this means the app size grows by a lot quickly when implementing this due to me needing a full copy of the model within the FlorisBoard app.

**Reply by ray-holmium** [reply]
I'm just one voice here, but you have my encouragement to try wild stuff like bundling a huge offline machine learning model (perhaps there are some other interesting things to do with the model down the road too, like translation). 

It defintely makes sense that you have the NLP in its own repo. If you define the NLP as a separate optional dependency for the time being, people can choose to install core florisboard without suggestions, and folks with spare processing power can choose to add the NLP package when ready as well.  Hopefully this can free you up from some of the concerns about how NLP affects the aims of the project as a whole. That can be a worry for a much later version, if you decide to integrate the packages. Because there's sort of a conflict between the need to have the core keyboard lightweight/accessible to as many users as possible and having an offline NLP. I don't think there's any problem just dividing and conquering for the forseeable future, rather than feeling like the project might be falling victim to feature creep.

I'm not an experienced dev but I am teaching myself right now and I think I can help squash some of the minor bugs with the core app. Will make sure not to flood you with arduous pull requests. Hopefully lots of others reading will join me in that. I'd like to think we as a community have your back on that so that feeling of having 5 things to do at once will let up a bit! 

I'm fairly new to contributing so hopefully haven't overstepped my bounds here. Just looking at the NLP problem from sort of an organizational framing perspective rather than a technical one. You're basically on the right path, and now is definitely the time to try a few wacky tricks or hail marys to see what sticks. Agree with OP 100% on the importance of this project to the overall ecosystem!

**Reply by pierre-eliep-met** [reply]
Thanks for sharing that also, I hope you will be able to work it out (I'm sorry no nothing of Kotlin, I wouldn't be able to help...), but my impression is that if word proposition does not work for now, we should just let it aside and work on some other parts...

**Reply by Markus40** [reply]
You could consider using an llm for word prediction. It should be possible to run a very small llm on the smartphone CPU. An llm should also give pretty good prediction results, even if it is so small.

The Pythia 70M, for example, can generate about 20 tokens per second on the CPU (assuming the phone is twice as fast as a RaspberryPi 4), or about 15 words per second, meaning you would get at least 3 words every 200ms. The llm gives you the probability of each next token, so you can probably get more out of it.

My hope is that llama.cpp will have Pythia support soon, then it should be easy to implement.

One drawback is support for other languages. Even if the model is trained mainly on English data, it will still work in other languages, just not as well.

**Reply by KaKi87** [reply]
Could you please just hold NLP and release bugfixes ? Thanks

**Reply by Markus40** [reply]
Instead of implementing a complicated word prediction engine, you could use something simple for the next release.

The Oxford English Dictionary contains about 300,000 main entries. With an average length of 5 characters and 3 possible next words, it would take about 7MB.
Then, for the 500 most common words, you could use 500 word pairs and predict 2 of the next words. This would take 6MB.

This would often give you 3-5 reasonable predictions, be fairly easy to implement and very light on the CPU.

More complicated algorithms could always be implemented in the future.

**Comment by ghost** [comment]
I don’t understand anything in development and what I’ll offer now, but maybe it will come in handy for you. Do not abandon your project and this will be the best keyboard. Perhaps if it’s not possible to implement proposals now, maybe you shouldn’t focus only on them

https://github.com/Pythagora-io/gpt-pilot

**Comment by ghost** [comment]
@patrickgold 

My two cents, dump suggestions! Focus on actual usable things that the user can immediately use that are within your scope of accomplishing. 

Gboard suggestions suck IMO. Swype was decent years ago. 90% of the time on Gboard the next word suggestion isn't even in the ballpark for me. 

I would much rather have a keyboard that is accurate than guessing what I want to say — Gboard swipe typing is still behind Swype from years ago. 

Functional things like a good editing keyboard, blacklist words, increase word priority, integration of sayboard with a good punctuation keyboard would all be infinitely better than suggestions IMO. 

If doing suggestions has put you this much off the project then it needs to be shelved! Suggestions are NOT a necessary feature.

**Comment by M00NJ** [comment]
I have some suggestions, ideas and advice as well.

@patrickgold you said that you are stuck in a loop. You can make the suggestions useful, but this will make them too slow or you can try to optimize performance, which will make them less useful again. It's a loose/loose situation. And the longer you are stuck in this loop, the more demotivated you will become. So my first piece of advice is to take a step back. Work on something else that furthers this project. Implement some open PR and maybe do a small release. To see progress happening after all this time of being stuck in the same place will bring back the motivation you need. Sometimes in life we need to take a step back to move forward. 

I also have an idea to fix the problem as well. You said you lack the expertise on statistical NLP and that there is information available online, but nothing really connects all the dots. That sounds like we need someone who has that expertise. One idea would be to announce that project is looking for donations to implement NLP and use these donations to hire someone who has the experience that is needed. 

On the topic of time management. You already rightfully stated, that this project has gotten to big for one person to maintain. So I suggest giving write access to people who are experienced and contribute frequently. Also having a group chat with 3-5 people working on the same project takes a lot of weight off your back. Being part of a team helps to stay motivated and putting your heads together on some issues will make solving them a lot faster. I'm part of the you-apps team and can confirm this. Bynro moving the apps to a separate organization and inviting me and another person so help maintain them has brought a lot of progress to these projects.

I hope some of what I said can be helpful to you. There's a reason why there are so few independently developed open source keyboard apps out there. Don't let this development plateau bring you down. Florisboard is a one-of-a-kind project and you can be very proud of what you have accomplished so far.

**Reply by Waelwindows** [reply]
I wholeheartedly agree with this sentiment. I've been anticipating this since #1109 (the early v0.3.13 days, 2 years ago) as the code base has gotten so big that it's hard for any **one** person to work on everything. As such, it is pertinent that we have different maintainers over many areas of the project, i.e. layouts, theming, NLP, etc. I believe that FlorisBoard has garnered enough of a community that it could maintain itself. Especially considering the state of the FLOSS Android Keyboards space.

I would like to nominate myself to be one of the maintainers of FlorisBoard as I'm still daily driving this keyboard. However, me being absent as I am (See #2054 and co) I don't think i would be any better than patrickgold himself 😅. What I _could_ promise is that I would try to work on the features I have said I'd work on, and help maintain/mentor any people in adjacent areas.

On a lighter note, I have recently taken a somewhat comprehensive course on AI which included NLP. I could look into perhaps implementing an AI NLP solution for FlorisBoard. However, as patrickgold said, that would require bundling in a copy of runtime such as TensorFlow Lite which would increase the bundle size by a lot, much more than what #2054 did.

Since I've been absent for a while, I don't have a clear view of what _has_ happened in the project yet. However, I believe the biggest roadblock currently is the extension system. it's a blocker for many stuff such as proper CJK input support, heavy duty NLP, and any work based that has an incompatible license with FlorisBoard's. GBoard itself isn't 150MB for no reason after all. It would also help ease of maintenance burden off of patrick as individual components could be developed separately.

Also, it is important that users realize all of this work is done on a **volunteer basis**.  AFAIK patrickgold is _still_ a university student which, in and of itself, commands so much time, let alone all of the other personal commitments, that I was surprised that patrick managed to _do_ this much. 

TLDR: @patrickgold Take a break, perhaps focus on other stuff but **please** delegate work to people who are _willing_ to work. We would be saddened to see this project go away when it has so much potential.

**Reply by patrickgold** [reply]
Thanks to both of you for your kind words and interesting ideas/proposals, that means a lot to me! It is true that the project has kinda grown over my head and me having more responsibilities than ever in my private life really increased my struggle with spending time for FlorisBoard, especially as getting NLP to run smoothly is a big demotivator as I am just completely stuck here. I will discuss the most important points below:

1. Suggestions/NLP: One of the key features I promised for 0.4 was suggestions, and as I don't like to break promises that was a big reason for me to keep trying so hard to implement this feature. However seeing me just being completely stuck I think I really need to take 2 steps back here and re-evaluate if I need to delay the feature to 0.5 (and maybe rewrite the NLP part from the ground up or using a different approach than statistical models). Not really sure here yet and I have to put more thoughts on this matter before making a decision, but seeing the desolate state of the project atm I think this could drastically improve the overall health of the project, as other bugs/PRs would finally get addressed by me.

2. Delegate work to volunteers: This could very well be an important step for this project to get back on track, as it would allow both me to focus on some features while other topics are generally managed by others. I have kinda avoided this until now because managing a team means more administrative work, but if I look at the project's state I think the administrative work is a lot less than the work that could be done by volunteers, so I think I should start to work on setting this up before anything else. Also I have to agree with you that having a team behind me is definitely a motivator for me, and it adds redundancy and stability to the project overall.

3. Extension/plugin system: I heavily agree with you here that a stable and more feature-rich extension system is required. Somewhere in the NLP part 2 PR I had the idea to experiment with plugins and services (in addition to extensions, not replacing), which are kinda UI-less apps that communicate with the core FlorisBoard app and are a lot more flexible, but as I tried to focus on getting the NLP core to run well I kinda got really stuck here and until now the plugin system idea is not fully developed yet. However I think before solving this problem a lot of other stuff needs to be addressed, and the first thing is getting the project back on track.

@Waelwindows As you mentioned that you would nominate yourself to be a maintainer of FlorisBoard I would gladly accept it! Having worked with you in the past you would be a great first addition to the maintainer team :)

So in conclusion: Getting the project back on track and assembling a maintainer team is the main priority now above anything else, which could possibly mean delaying the suggestions to 0.5. I will share more details soon in a new announcement and new project roadmap.

**Reply by GrimPixel** [reply]
I have a question: why is it so big?
I doubt if there are too much complexity already there. For example, to create a layout, I have to find the XML values of characters and input them. This is uniquely complicated among keyboard programs. I believe that problems of complexity exist elsewhere, too. When I said it on Matrix.org, the response was an imagination of creating something more complicated: a layout editor. 
What needs to be done is to inspect the old code and focus on the optimisation, to maintain simplicity. I see the lack of interactions between the developer and users, as many issues unanswered. We need to come up with all different solutions of a problem and discuss them. Good documentation is helpful with understanding what the code looks like overall, as well as being a key to gaining people's interests.

**Comment by ghost** [comment]
https://github.com/languagetool-org/languagetool

**Comment by Kreuger** [comment]
Thanks for the post OP and all of the ongoing discussion. I may not be able to code but still have all 3 builds installed and am happy to test anything out. I think from the sounds of it, Patrick is getting back on track in the right way.

**Comment by kpalang** [comment]
Even though I might not have the capacity to help maintaining FlorisBoard, I still want to cheer Patrick on to continue working on it and have him know that people really like his creation.
I use Floris exclusively on all my Android devices and would really love to see it move forward.

**Comment by GrahamboJangles** [comment]
I second the use of LLM extensions to fix the suggestions, since these models are quite light now with LLaMA derivative C++ implementations, and now with most phones having lots of storage space I'd have no problem storing a language model on my phone (especially if I can put it on an SD card). Like @Charles7z said, even the Gboard suggestions are pretty bad so it should not be hard to beat even with a simple 7b parameter 4-bit quantized LLM.

The biggest issue I foresee with this is RAM. This Android implementation says it requires 8gb of RAM https://github.com/ggerganov/llama.cpp/discussions/750 which is not so easy to come by, while also considering the requirements for the rest of the programs run by Android. Maybe the RAM requirements could be drastically reduced by reducing the context length size down to a sentence or two, since you're only trying to predict one word. 

Also just a thought, is funding is not a issue, then it sounds like the funding could be reinvested into the project by hiring programmers.
